code/website/site.hs

```diff
@@ -33,46 +33,101 @@ getExtensionName _ = error "Expected some extension."
+    \y -> if y then map (getSrc repoRoot) . lines . rstrip <$> readFile (path ++ x ++ "/src") else return []) names
+    
 
   -- a list of Just descriptions or Nothing based on existence and contents of desc file.
   descriptions <- mapM (\x -> doesFileExist ("descriptions/" ++ x ++ ".txt") >>=
-      \y -> if y then Just . rstrip <$> readFile ("descriptions/" ++ x ++ ".txt") else return Nothing) names
+    \y -> if y then Just . rstrip <$> readFile ("descriptions/" ++ x ++ ".txt") else return Nothing) names
 
   -- creates a list of SRSVariants, so a list of list of tuples.
   -- the outer list has an element for each example.
   srss <- mapM (\x -> sort . getSRS <$> listDirectory (srsPath path x srsDir)) names
 
   -- returns the IO list of examples with constructor E containing
-  -- the name of the example, sources (if applicable), and the list of Variants.
+  -- the name of the example, sources, list of variants, and description
   return $ map (\(name, source, srs, desc) -> E name source srs desc) $ zip4 names sources srss descriptions
 
 maybeField :: String -> (Item a -> Compiler (Maybe String)) -> Context a
 maybeField s f = Context $ \k _ i -> do
+  -- val will be of type Maybe String
   val <- f i
+  -- if val is Just and the input string matches the html file paramater,
+  -- then we use the String contained in val
   if k == s && isJust val then
     return $ StringField $ fromJust val
   else
+    -- Otherwise, fail acts to not use the $if$ parameter in the html file
     fail $ "maybeField " ++ s ++ " used when really Nothing. Wrap in `$if(" ++ s ++ ")$` block."
 
+-- List if unempty
+maybeListFieldWith :: String -> Context a -> (Item b -> [c]) -> (Item b -> Compiler [Item a]) -> Context b
+-- takes as input: the string to be replaced, the context to be used as is,
+-- a function to be used on input to return a list to check for emptiness,
+-- and the function to be used regularly on input
+maybeListFieldWith s contxt checkNull f = listFieldWith s contxt
+  -- If the list is empty, fail is used so as not to create a specific section
+  (\x -> if null (checkNull x) then fail "No code files for example" else f x)
+
 mkExampleCtx :: FilePath -> FilePath -> Context Example
 mkExampleCtx exampleDir srsDir =
+  -- each function is applied to the item containing an Example of form:
+  -- E Name [CodeSource] SRSVariants Description
+
+  -- here, listFieldWith used with multiple versions of SRS per example
   listFieldWith "srs" (
+    -- field will replace the string parameters in the html with the returned string
+    -- <> acts as a monad concatenator
+
+    -- returns filename from ((filename, TYPE), E _ _ _ _)
     field "filename" (return . fst . srsVar) <>
+    -- returns TYPE from ((_, _), E _ _ _ _)
     field "type" (return . snd . srsVar) <>
+    -- returns name from ((_, _), E name [codeSource] srsVariants description)
```

Comment:
I would add here (and really should have when I wrote it, the `name` we define here is the same `name` as the outer scope, but for some reason the templating language doesn't keep the outer variables in scope when we enter a for loop, so we redefine it to make it available there.